### PR TITLE
fix: a 50x should be reported as invocation failure

### PR DESCRIPTION
**Pull request by dominique-pfister**
When this lambda is invoked by SQS, it should *throw* an error to indicate a recoverable failure. At the moment, we just see a message in the log, e.g.:
```
Fetching https://spark-website-adobe.hlx.page/tw/express/create/flyer/grand-opening.html failed: statusCode: 504, message: '<!doctype html>
<html lang="en">
<head>
<title>Gateway timeout</title>
```

See here: https://helix.coralogix.com/#/query-new/logs?id=2VhZw7ryYbK

### CODE REVIEW

File: src/index-pipelines.js [hunk 1]
```diff
@@ -92,14 +92,14 @@ async function fetchHTML(url, log) {
...
+    const snippet = body < 100 ? body : `${body.substr(0, 100)}...`;
+    const message = `Fetching ${url.href} failed: statusCode: ${resp.status}, body: '${snippet}'`;
+    return { error: { path: url.pathname, status: resp.status, message } };
   }
   const s = body.trim();
   if (s.substring(s.length - 7).toLowerCase() !== '</html>') {
-    log.warn(`Document returned from ${url.href} seems incomplete (html end tag not found)`);
-    return { error: { path: url.pathname, status: 500, reason: 'document incomplete' } };
+    const message = `Document returned from ${url.href} seems incomplete (html end tag not found)`;
+    return { error: { path: url.pathname, status: 500, message } };
```
Comment (dominique-pfister) [inline src/index-pipelines.js:102]: incomplete returns 500 (it might recover if the ESI failed)

File: src/index-pipelines.js [hunk 2]
```diff
@@ -92,14 +92,14 @@ async function fetchHTML(url, log) {
     body = e.message;
   }
   if (!resp.ok) {
-    const message = body < 100 ? body : `${body.substr(0, 100)}...`;
-    log.warn(`Fetching ${url.href} failed: statusCode: ${resp.status}, message: '${message}'`);
-    return { error: { path: url.pathname, status: resp.status, reason: message } };
+    const snippet = body < 100 ? body : `${body.substr(0, 100)}...`;
+    const message = `Fetching ${url.href} failed: statusCode: ${resp.status}, body: '${snippet}'`;
+    return { error: { path: url.pathname, status: resp.status, message } };
```
Comment (dominique-pfister) [inline src/index-pipelines.js:97]: this is where the 504 might be generated

File: src/index.js [hunk 3]
```diff
@@ -141,7 +141,7 @@ async function handleDelete({ config, handler }, change, log) {
   if (!handler) {
     return {
       status: 400,
-      reason: 'Handler not available, parameters missing or target unsuitable',
+      message: 'Handler not available, parameters missing or target unsuitable',
```
Comment (dominique-pfister) [inline src/index.js:144]: I create an `Error` in the main function, so for sake of simplicity, I renamed `reason` to `message` everywhere.

File: src/index.js [hunk 4]
```diff
@@ -359,11 +359,15 @@ async function main(req, context) {
...
-  const status = results.reduce(
-    (curr, r) => (r.status >= 500 ? r.status : curr), 207,
-  );
+  const failure = results.find((r) => r.status >= 500);
+  if (failure) {
+    const e = new Error(failure.message);
+    e.status = failure.status;
+    e.path = failure.path;
+    throw e;
+  }
```
Comment (dominique-pfister) [inline src/index.js:368]: If there was a recoverable failure, throw that instead of returning it as a `Response`

File: src/index.js [hunk 5]
```diff
@@ -372,5 +376,4 @@ module.exports.main = wrap(main)
   .with(bodyData)
   .with(recordsWrap)
   .with(helixStatus)
-  .with(logger.trace)
```
Comment (dominique-pfister) [inline src/index.js:375]: This has to be removed because it will unconditionally catch errors thrown by the Lambda

File: test/index.test.js [hunk 6]
```diff
@@ -122,10 +131,14 @@ describe('Index Tests', () => {
...
-          server: 'nock',
-        }];
+        const metadata = `${path}.json`;
+        let status = 200;
+        let headers = { 'last-modified': 'Mon, 22 Feb 2021 15:28:00 GMT' };
+
+        if (fse.existsSync(metadata)) {
+          ({ status, headers } = fse.readJSONSync(metadata));
+        }
+        return [status, fse.readFileSync(path, 'utf-8'), headers];
```
Comment (dominique-pfister) [inline test/index.test.js:141]: When asking for a HTML, allow returning metadata as well in file with added extension `.json`

File: src/index.js [hunk 7]
```diff
@@ -241,12 +241,12 @@ async function handleUpdate({
       if (!record.sourceHash) {
         const message = `Unable to update ${path}: sourceHash in indexed document is empty.`;
         log.warn(message);
-        return mapResult.error(path, message);
+        return mapResult.error(400, path, message);
       }
```
Comment (dominique-pfister) [inline src/index.js:245]: Changed to status code 400, as we can't recover from a bad document

File: src/index.js [hunk 8]
```diff
@@ -241,12 +241,12 @@ async function handleUpdate({
...
       if (isOutdated(record, headers, change)) {
         const message = `Unable to update ${path}: indexed record is outdated.`;
         log.warn(message);
-        return mapResult.error(path, message);
+        return mapResult.error(409, path, message);
       }
```
Comment (dominique-pfister) [inline src/index.js:250]: When the record is outdated, we have a conflict (a newer version of that document has already been indexed)

File: src/utils.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * - the change UID and the HTML UID (or source hash) are different
+ * - the change event time is later than the last modified of the HTML
+ *
+ * In that case, the change reported relates to a different and more recent
+ * item, so we shouldn't add an index record for an outdated item.
+ *
+ * @returns true if the record is outdated and shouldn't be used for indexing
+ */
+function isOutdated(record, headers, change) {
```
Comment (dominique-pfister) [inline src/utils.js:27]: Refactored isOutdated to create a test case

**Review by tripodsan — APPROVED** [review]

**Pull request merged**